cosigner/authstatestores.go

```diff
@@ -0,0 +1,168 @@
+package cosigner
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type AuthStateStore interface {
+	CreateNewAuthSession(pkt *pktoken.PKToken, ruri string, nonce string) (string, error)
+	LookupAuthState(authID string) (AuthState, bool)
+	UpdateAuthState(authID string, authState AuthState) error
+	CreateAuthcode(authID string) (string, error)
+	RedeemAuthcode(authcode string) (AuthState, string, error)
+}
+
+type AuthStateInMemoryStore struct {
+	AuthIDIssuer     *AuthIDIssuer
+	AuthStateMap     map[string]*AuthState
+	AuthCodeMap      map[string]string
+	AuthStateMapLock sync.RWMutex
+	AuthcodeMapLock  sync.RWMutex
+}
+
+func NewAuthStateInMemoryStore(hmacKey []byte) *AuthStateInMemoryStore {
+	return &AuthStateInMemoryStore{
+		AuthStateMap:     make(map[string]*AuthState),
+		AuthCodeMap:      make(map[string]string),
+		AuthcodeMapLock:  sync.RWMutex{},
+		AuthStateMapLock: sync.RWMutex{},
+		AuthIDIssuer:     NewAuthIDIssuer(hmacKey),
+	}
+}
+
+func (s *AuthStateInMemoryStore) LookupAuthState(authID string) (AuthState, bool) {
+	s.AuthStateMapLock.RLock()
+	as, ok := s.AuthStateMap[authID]
+	s.AuthStateMapLock.RUnlock()
+	return *as, ok // Pass by value to prevent writes to the original
```

Comment:
I think I get what you're trying to achieve by making `LookupAuthState` and `RedeemAuthcode` return values instead of pointers (to make a copy and prevent writes to the original value backed by the `map`), but since `AuthState` has a field that is a pointer (`Pkt`), it's not truly a copy (only the fields not backed by pointers are copied). This is really a shallow copy behind the scenes--the returned `AuthState` will have a copy of the `Pkt` pointer to the original `Pkt` of the `AuthState` stored in the `map`.

Reply:
This has been fixed